### ISSUE TITLE
SMW::Property::initProperties called on every load

### ISSUE THREAD

**Issue by YOUR1** [SemanticMediaWiki/SemanticMediaWiki#4973]
Hi,

We have a hook that registers special properties through the `SMW::Property::initProperties` hook. Our hook also sets default properties. 

When debugging I found out that the `SMW::Property::initProperties` is loaded on every hit, and I was wondering if this is default behaviour?

Having the initProperties loaded on every page load gives us additional load, due the fact we want to set default values. 

The code somewhat looks like (its just an example of how things work):
```php
public static function initializeSpecialProperties( \SMW\PropertyRegistry $propertyRegistryInstance ) {
    $store = smwfGetStore();
    $allPropertyDefinitions = FunctionToGetPropertyDefinitions();
    // Loop over all custom property defs
    foreach ($allPropertyDefinitions as $propertyDefinition) {
        $id = '___'.$propertyDefinition[0];
        $type = $propertyDefinition[1];
        $label = $propertyDefinition[2];
        $values = isset($propertyDefinition[3]) ? $propertyDefinition[3] : false;
        $visibility = true;
        $annotatable = false;
        // Register the actual property
        $propertyRegistryInstance->registerProperty($id, $type, $label, $visibility, $annotatable);
        // For properties that have values 
        // This is actually bad since this will be run on every
        // pageload. 
        if ( $values ) {
            $allowsValueProperty = new SMWDIProperty('_PVAL');
            $typeProperty = new SMWDIProperty($id);
            $typePropertyPage = $typeProperty->getDiWikiPage();
            $data = new SMWSemanticData($typePropertyPage);
            foreach ($propertyDefinition[3] as $allowedValue) {
                $dataItem = new SMWDIBlob($allowedValue);
                $data->addPropertyObjectValue($allowsValueProperty, $dataItem);
            }
            $store->updateData($data);
        }
    }
}
```

The part where the default values are added to the special property seems very inefficient and I was wondering what optimisations can be done.

I'm using MW1.31 & SMW3.1 on PHP7.3 btw.

**Comment by JeroenDeDauw** [comment]
If I remember correctly, the special properties added by these hooks are not stored in the database. Instead they are just in-memory info. So it's not that big of a deal that the hook gets called on every hit.

I agree that doing a write on every hit, like your code is doing, is not ideal. If these allowed values are fixed, then you could add them as a schema update, using the MediaWiki hook for that. Then it will only get executed once when update.php is run.

**Comment by YOUR1** [comment]
> If I remember correctly, the special properties added by these hooks are not stored in the database. Instead they are just in-memory info. So it's not that big of a deal that the hook gets called on every hit.
Yeah, thats what I found out, a simple array key/value store. Thanks for confirming. 
 
> I agree that doing a write on every hit, like your code is doing, is not ideal. If these allowed values are fixed, then you could add them as a schema update, using the MediaWiki hook for that. Then it will only get executed once when update.php is run.
That's the same suggestion I had in mind. But since they are actually dumb things in an array store I was hoping that I could bypass writing anything to the DB since they are in fact simply stored in the array.

**Comment by JeroenDeDauw** [comment]
If you find a way to add the allowed property info to the right SMW servcice(s), then I suspect you indeed won't need to write to the DB. Doesn't `$propertyRegistryInstance` have anything for this?

**Comment by YOUR1** [comment]
@mwjames do you have any suggestions?

**Comment by YOUR1** [comment]
I moved the if(values) part to `SMW::SQLStore::Installer::AfterCreateTablesComplete`, that seems to be working correctly.
